### ISSUE TITLE
ES5 transpilation: UpdatableTag.compute errors with nested tracked properties

### ISSUE THREAD

**Issue by christophersansone** [glimmerjs/glimmer.js#208]
Hi there!

I am upgrading a Glimmer app from 0.7 to 0.13, and I have come across the following error.  Unfortunately my project requires ES5 support, which is a factor here, as it otherwise works for ES6 compilation.  I have tested against 0.11.2, 0.13, and 0.14-alpha.12 with the same results, and it can be reproduced on a brand new Glimmer app when enabling ES5 output.

The gist: _when compiling to ES5, if a tracked property relies on another tracked property, all rendering and interaction will stop working without any sort of error._

Let me start by saying that if I thought this was some IE11 bug or something that seems not worth trying to fix, I wouldn't file the issue.  However, it seems like a minor thing that can be solved to continue making Glimmer ES5 compatible.

To reproduce:

1. `ember new es5issue -b @glimmer/blueprint`
2. Change the last line of `config/targets.js` to `module.exports = 'es5';`. Also reproducible by adding `IE 11` to the list of browsers.
3. Include `babel-polyfill` for the essentials.  You can `npm install babel-polyfill` and then add the following as the first line to `index.ts`: `import '../../node_modules/babel-polyfill/browser';`.  Also reproducible by just adding a `<src>` tag to `index.html` to pull it in from Babel's CDN.
4. In the main component, use the demo code from the Glimmer Playground as an example:

```
component.ts:

import Component, { tracked } from '@glimmer/component';
export default class extends Component {
  @tracked count = 1;
  increment() {
    this.count++;
  }
}

template.hbs:

<h1>Welcome to the Glimmer Playground!</h1>
<p>You have clicked the button {{count}} times.</p>
<button onclick={{action increment}}>Click</button>
```

If you run this, it works as expected.  However, if we simply add another tracked property that reads `count`, it will fail:

```
component.ts:

...
  @tracked
  get anotherCount() {
    return this.count; 
  }
...

template.hbs:

...
{{anotherCount}}
...
```

This should work, and it does work via ES6, but it fails here... and without any kind of console error or warning about what happened.

Chasing the rabbit down the hole a bit more, I looked through the renderer code and discovered that if I added the following to my `index.ts`, I could find out if the renderer failed:

```javascript
app._notifiers.push([ (() => console.log('success')), ((e) => console.log('error', e))]);
```

By doing this, I found a helpful error message that was getting swallowed:

```
error TypeError: this.tag.value is not a function
    at UpdatableTag.compute (app.js:4215)
    at UpdatableTag.value (app.js:4128)
    at app.js:3994
    at TagWrapper.value (app.js:3981)
```

In this case, it appears it is expecting `this.tag` to be an `UpdatableTag`, but instead it is a `Set` containing a single `TagWrapper` referencing an `UpdatableTag`.  It looks like the ES6 version has a `TagsPair` object which does have a `value` method, but the ES5 version is a `Set` that does not have a `value` method.

Currently, to fix this, I monkey-patched `UpdatableTag` in my project's `index.ts` by checking for a Set and (recursively) finding the maximum value:

```javascript
import { UpdatableTag } from '@glimmer/reference';
...
function tagValue(tag) {
  if (tag.value) {
    return tag.value();
  } else if (tag instanceof Set) {
    let value = -Infinity;
    tag.forEach((item) => {
      const v = tagValue(item);
      if (v > value) {
        value = v;
      }
    });
    return value;
  } else {
    throw new Error('Cannot determine tag value');
  }
}

UpdatableTag.prototype.compute = function() {
  return Math.max(this.lastUpdated, tagValue(this.tag));
};
```

It is not pretty, but it appears to solve the problem and seems to evaluate the same way as `TagsPair`.

**Comment by rwjblue** [comment]
I don't really understand why we differ between the ES5 and ES6+ builds, I would expect both to be using a `TagPair` here. Do you see why?

**Comment by christophersansone** [comment]
I'm not sure either, it seems like it's deep in `glimmer-vm`.  It's defined [here](https://github.com/glimmerjs/glimmer-vm/blob/v0.41.4/packages/@glimmer/reference/lib/validators.ts).

Looking at the actual distribution of `@glimmer/reference`, there are separate folders for `es5` and `es2017`, and the `validators.js` file does indeed contain the `TagsPair` class, so I'm not sure why it would choose a Set instead.  Maybe someone on the `glimmer-vm` team can offer some insight?

**Comment by rwjblue** [comment]
I'm wondering if this is just a case of an invalid `npm publish`? :thinking:

**Comment by christophersansone** [comment]
Is there any way I can easily test this theory?  Maybe change `package.json` to point the glimmer packages to the Github repo instead?  Or would there probably still be deeper dependency issues?

**Comment by christophersansone** [comment]
Thanks for the help, @rwjblue... really appreciate it.  I managed to dig a bit deeper, and I think I found the root cause, and it makes sense why the ES5 transpilation has a problem.

Here is the Tracker class (source is [here](https://github.com/glimmerjs/glimmer.js/blob/6eaced5b337bb3cc00f6aa317ff2d8bd1534234b/packages/%40glimmer/component/src/tracked.ts#L8)):

```javascript
class Tracker {
  private tags = new Set<Tag>();

  add(tag: Tag) {
    this.tags.add(tag);
  }

  combine(): Tag {
    let { tags } = this;

    if (tags.size === 0) return CONSTANT_TAG;
    return combine([...tags]);
  }
}
```

I have confirmed that `tags` is the `Set` that is incorrectly being passed.  The problem appears to be with this line:

```javascript
return combine([...tags]);
```

Running `[...tags]` in the console returns a flattened array containing the elements of the `Set`.  This is transpiled into ES5 as:

```javascript
return combine([].concat(tags));
```

Running `[].concat(tags)` in the console returns an array with the `Set` as the single item in the array.  Bottom line is that the set is flattened into an array in ES6, but it is not flattened in ES5.

I'm thinking we can fix it by changing it to:

```javascript
return combine(Array.from(tags));
```

This is probably a bit clearer and safer, since it better guarantees that it will properly convert an array-like object into an array.  I've confirmed that changing this line in my `dist/app.js` solves it.  Of course, `Array.from` is not ES5, but polyfilling is a requirement anyway, so no worries there.

PR is on its way, I guess we can discuss the code change there.  Thanks again!
